Fix calculation and implementation of dial ratio and the max dialed connections limit

oneeman opened the pull request:
### Description

Fixes two behaviors in `p2p/Server` which cause it to get more dialed connections than it should based on the dial ratio (which in turn impairs the ability of new nodes joining the network to find peers with open incoming connection slots):

1. The bug described in #1669, which affects LES servers.
2. The fact that the dialer attempts two dials for every dialed peer connection it's missing: https://github.com/celo-org/celo-blockchain/blob/147571e5b5aab36b6de4d839a3ee9c1f85dbd94b/p2p/dial.go#L381

Initially I was planning to just fix #1669.  But then when I tested it out and saw that it still ends up with more dialed peers than it should, I remembered this `* 2` and realized that it would have to be removed in order to get the desired behavior.

This `* 2` was introduced upstream in https://github.com/ethereum/go-ethereum/pull/20592.  Prior to that, it would attempt one dial for every additional dialed peer that it wanted: https://github.com/ethereum/go-ethereum/blob/58cf5686eab9019cc01e202e846a6bbc70a3301d/p2p/dial.go#L149-L160

The new behavior was introduced in that PR, but was not mentioned in the PR description, commented on in the code, nor commented on during PR review.  But, seeing as it impairs the functioning of the dial ratio and geth was doing fine without it up to 1.9.10, I don't see a problem with removing it.  It means that you will be slower to dial peers initially (though there is at any rate the `maxActiveDials` limit which defaults to 50) and if a dialed peer connection drops you will try to refill it one at a time instead of two at a time (if that makes sense), but that seems reasonable.

### Other changes

Updated a test to work with the new behavior (the test assumed two dial slots for every dialed peer we want).

### Tested

Verified that the dial ratio is respected by connecting a node with low limits to Alfajores and seeing that it only connects to  the right number of dials peers, whether or not it's running the LES server.

### Related issues

- Fixes #1669

trianglesphere commented on p2p/dial_test.go:
```diff
@@ -58,61 +58,55 @@ func TestDialSchedDynDial(t *testing.T) {
-				newNode(uintID(0x05), "127.0.0.1:30303"),
-				newNode(uintID(0x06), "127.0.0.1:30303"), // not dialed because there are only two slots
+				newNode(uintID(0x05), "127.0.0.1:30303"), // not dialed because there are only two slots
+				newNode(uintID(0x06), "127.0.0.1:30303"), // ..
```
```suggestion
				newNode(uintID(0x06), "127.0.0.1:30303"), // ...
```

oneeman replied:
Fixed in new commit

trianglesphere commented on p2p/dial_test.go:
```diff
@@ -58,61 +58,55 @@ func TestDialSchedDynDial(t *testing.T) {
 				newNode(uintID(0x04), "127.0.0.1:30303"),
-				newNode(uintID(0x05), "127.0.0.1:30303"),
-				newNode(uintID(0x06), "127.0.0.1:30303"), // not dialed because there are only two slots
+				newNode(uintID(0x05), "127.0.0.1:30303"), // not dialed because there are only two slots
```
not dialed b/c 1 slot?

oneeman replied:
Fixed in new commit

trianglesphere reviewed (approved):
I've got a couple small comments on the tests, but otherwise it looks good to me.

oneeman commented:
> I've got a couple small comments on the tests, but otherwise it looks good to me.

Thanks!  Updated those comments as well as another one in a new commit, and then I rebased on master.
